### ISSUE TITLE
How to improve Lookup filter quality?

### ISSUE THREAD

**Issue by GustasBr** [MetalPetal/MetalPetal#162]
Hey,

I'm wondering how I could improve filters. Basically, I'm creating filters using Photoshop or Lightroom. Then I export it to color lookup table. I'm using these export parameters:

- Color space: sRGB (Bit depth: 8 bits per component)
- Image sizing: 72 pixels/inch.
- File format PNG (depth: 8 or 16)

My code:
```
guard let lookupimage1 = UIImage(named: "lookup_\(effectProperties.lookup_id).png") else { return }`

guard let cgImage1 = lookupimage1.cgImage else { return 
let filter1 = MTIColorLookupFilter()

filter1.inputImage = outputImage
filter1.inputColorLookupTable = MTIImage(cgImage: cgImage1, options: [MTKTextureLoader.Option.SRGB: false], alphaType: .alphaIsOne)
filter1.intensity = effectProperties.intensity
outputImage = filter1.outputImage
```

This functionality works pretty well but I really need more quality. When I'm applying those filters on Adobe products it looks very well and the image quality is the best. When I'm using MetalPetal I get a little bit worse quality and sometimes really weird color glitch (look at the sky). Of course other color lookup tables works better than this one. But still I can see really easily that the quality and the colors are so much different when comparing to photos edited on Adobe.:
![IMG_8124](https://user-images.githubusercontent.com/18195248/85186535-50e74000-b2a2-11ea-9f05-34bedfd47ddf.JPG)
 I can add more examples if that would help.

So is it possible to do something different to achieve better quality? Maybe my code, export parameters or anything else is bad. 

Lookup Table:
![lookup_2](https://user-images.githubusercontent.com/18195248/85200134-7a818500-b2fd-11ea-8389-e49aa113422a.png)

Please, help me.
Thank you

### Checklist
- [x] I've read the [README](https://github.com/MetalPetal/MetalPetal/blob/master/README.md)
- [x] I've searched for [existing GitHub issues](https://github.com/MetalPetal/MetalPetal/issues)

**Comment by YuAo** [comment]
A PSD file of the filter would be helpful.

**Comment by GustasBr** [comment]
> A PSD file of the filter would be helpful.

Thank you for your fast reply. After investigating more I found that the problem can occur because I'm using Magic Export Tool on Adobe Lightroom which doesn't export Highlights, Shadow, Clarity, Sharpness, Noise Reduction & other parameters. OR maybe it's even impossible when using color lookup tables (i think that's more likely).. So, what I decided to do is to make those changes with MetalPetal other functionality. I'm started using clarity as MPS Unsharp Mask, Noise reduction as MTISurfaceBlurFilter (it sucks that it takes forever to apply this SurfaceBlur filter..) . Sadly, shadows the only function which is not working very well. Somehow the colors become really weird and glitchy (similar like with my sky photo above). Anyway, the result when applying those parameters is now better. Even though I would like my photos to be exact like in Lightroom but meh. It's what it's. 

To sum up I have two final questions for you:

-  Why sometimes I get that really weird color glitch like in the photo above? (also, you can find lookup table now in my edited post). I think I get that glitch because my filter changes all of the colors sooo much. But anyway, with Lightroom I don't get that kind of thing. Also, this problem really often occurs with photos where you can find human face. Because of the skin.
- Second question would be - if it's possible to make lookup table with all of the parameters mentioned before. (Highlights, Shadows, Clarity, Noise Reduction & etc)?.

Thank you.

**Comment by YuAo** [comment]
That's a lot of information. And I'm busy with work currently. I think I could get some time to investigate and answer your questions later this week.

**Comment by YuAo** [comment]
It looks like there is something wrong with your lookup image. A lot of artifacts along the edges of each color block. I think this is what causes the color glitch. You'll need to investigate why these artifacts appear.
<img width="664" alt="Screen Shot 2020-06-28 at 12 21 53 PM" src="https://user-images.githubusercontent.com/1234944/85937861-0e74d180-b93a-11ea-8d66-8f80e7df8608.png">

For your second question. No. Color lookup tables is just direct color map. Highlights, Shadows, Clarity, Noise Reduction, Blur, Smoothing these can all be local color operations. That means an output pixel of these operations is often related to its neighborhood pixels, with different neighborhood pixels one color with the same operation parameter can produce different outputs. It's hard to emulate Photoshop's shadows, highlights, clarity, etc - these are all private algorithms.
